### DISCUSSION TITLE
How I can (partially) disable sorting/ordering in `arbitrary_source_item_ordering`?

### DISCUSSION THREAD

**Discussion by TimDiekmann** [rust-lang/rust-clippy#13675] [Q&A]
Very recently the new [`arbitrary_source_item_ordering`](https://rust-lang.github.io/rust-clippy/master/#arbitrary_source_item_ordering) was added. It is a lint with quite a few configurations, but I'm struggling a bit how I can disable alphabetical sorting.

I'm particularly interested in enforcing some grouping, i.e. group `use`, `mod`, etc., but I don't get it to not enforce alphabetical item ordering at the same time. I also did not manage to only enforce some groupings but not all at the same time (e.g. to not group `lower_snake_case` items but the other groups).

Could you give me some pointer how a configuration for this might look like?

**Comment by flip1995** [comment]
cc @decryphe (the author)

**Comment by decryphe** [comment]
Currently it's not possible to disable alphabetical ordering within groups - I've already put that on my todo-list and will add it as a feature in the coming week or two (should be a very small change).

Enforcing some groupings but not others can be achieved by declaring the groups you want, and putting everything else in a catch-all group.

That could look a bit like this, adjusted as per your needs:
```
module-item-order-groupings = [
    [ "modules", [ "extern_crate", "mod", "foreign_mod" ], ],
    [ "use", [ "use", ], ],
    [ "everything_else", [ "macro", "global_asm", "static", "const", "ty_alias", "enum", "struct", "union", "trait", "trait_alias", "impl", "fn", ], ],
]
```

Until I add the option to switch off alphabetical sorting within groups, it's probably not that useful though.

Now that I think of it, it could work better with a catch-all for that, I could add the option to put an asterisk in one of the groups to make configuration easier.

**Reply by TimDiekmann** [reply]
Amazing, thank you!

Also, the example of the grouping instruction is indeed helpful. I Misread the documentation and thought the first item in the inner arrays are predefined. 
Would you mind posting here when you added the functionality? Thanks for your work 🙂

**Reply by decryphe** [reply]
That's a good hint for the documentation, I'll add it with the update.

I'll make sure to ping you here once the feature is added.

**Reply by TimDiekmann** [reply]
> Now that I think of it, it could work better with a catch-all for that, I could add the option to put an asterisk in one of the groups to make configuration easier.

I think an asterisk makes sense, yes. I don't know what you're planing on configuring the groups, but some way to "just leave this group alone" would also be quite helpful, e.g. either by leaving out some items (such as `enum` or `union` or make this explicitly configurable.

**Reply by TimDiekmann** [reply]
> That's a good hint for the documentation, I'll add it with the update.
> 
> I'll make sure to ping you here once the feature is added.

Great, thank you!

**Reply by decryphe** [reply]
> I think an asterisk makes sense, yes. I don't know what you're planing on configuring the groups, but some way to "just leave this group alone" would also be quite helpful, e.g. either by leaving out some items (such as `enum` or `union` or make this explicitly configurable.

The crucial part here is to ensure that it's deterministic where the "leave them alone" items should be placed. As I basically scan a module from top-to-bottom, it currently needs to know which order should apply for all items (a bit like the difference between Eq and PartialEq).
The exact behaviour can still be defined though:
- Leave-alone items (e.g. not mentioned in the config) can be simply ignored anywhere in the file (so a spurious `fn` can be placed in-between `use` statements), or
- Leave-alone items can be put in the catch-all group. Here it's still open if the catch-all group should order within or not, or how exactly it should behave.

For the config switch to turn on or off alphabetical ordering within groups, I think changing the default to off would be sensible, as the feature most people seem to be interested in is the configurable groupings.

**Reply by TimDiekmann** [reply]
> The crucial part here is to ensure that it's deterministic where the "leave them alone" items should be placed. 

Makes sense, yes! I like the optionality to leave them out or put them in a catch-all group (while then they strictly wouldn't be a leave-alone item anymore). The only real issue I see on leaving them out is that maybe a custom configuration may miss new item kinds. But maybe that's also a good approach because it's more future proof if more group items are supported (if any).

> I think changing the default to off would be sensible, as the feature most people seem to be interested in is the configurable groupings.

I think so, yes, because I'd imagine that sorting alphabetically is most interesting for `use` statements, which can be done by `rustfmt` anyway. In almost all other locations at least I try to group by semantic meaning (e.g. `fn new`, `fn from_x`, `fn from_y`) which is not possible to enforce through lints. You could add more `impls` blocks but in the above example alphabetical sorting would be in the wrong order as well (at least it wouldn't be how I would expect them to be laid out).

**Reply by decryphe** [reply]
@TimDiekmann I've created a PR that adds this config option: #13718

**Reply by decryphe** [reply]
@TimDiekmann I've updated the PR, feel free to have a look at it, it should be possible to configure the lint for your use-case now using the new config option `module-items-ordered-within-groupings`.

**Reply by TimDiekmann** [reply]
Great to hear and looking good, thank you! Will definitely try it out 🚀
